multitenant: prevent tenant from overriding next tenant's span config

ecwall opened the pull request:
Fixes #95882

We want to ensure we have a split at the non-inclusive end of the tenant's
keyspace, which also happens to be the inclusive start of the next
tenant's (with ID=ours+1). If we didn't do anything here, and we were the
tenant with the highest ID thus far, our last range would extend to /Max.
If a tenant with a higher ID was created, when installing its initial span
config record, it would carve itself off (and our last range would only
extend to that next tenant's boundary), but this is a bit awkward for code
that wants to rely on the invariant that ranges for a tenant only extend
to the tenant's well-defined end key.

So how do we ensure this split at this tenant's non-inclusive end key?
Hard splits are controlled by the start keys on span config records[^1],
so we'll try insert one accordingly. But we cannot do this blindly. We
cannot assume that tenants are created in ID order, so it's possible that
the tenant with the next ID is already present + running. If so, it may
already have span config records that start at the key at which we want
to write a span config record for. Over-writing it blindly would be a
mistake -- there's no telling what config that next tenant has associated
for that span. So we'll do something simple -- we'll check transactionally
whether there's anything written already, and if so, do nothing. We
already have the split we need.

[^1]: See ComputeSplitKey in spanconfig.StoreReader.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96014)
<!-- Reviewable:end -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96014)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @knz)

___
*[`pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split_regression` line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMjCazZ-gpcX44JaBpV:-NMjCazZ-gpcX44JaBpW:bjgr9mg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fffb1bbe7ca04f5a4afc5ddecc415f92b3d453/pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split_regression#L119)):*
> ```
> ...
> /Tenant/10{-\x00}                          database system (tenant)
> /Tenant/11{-/Table/4}                      database system (tenant)
> ```

This is now no longer overridden (compare vs 1st commit in PR).
___
*[`pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split_regression` line 127 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMjCszqCQmcExI3bHI2:-NMjCszqCQmcExI3bHI3:bmbqz34) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fffb1bbe7ca04f5a4afc5ddecc415f92b3d453/pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split_regression#L127)):*
> ```
> # And tenant=11, since it assumes its the only writer, simply does not observe
> # this mutation.
> mutations tenant=11
> ```

Is this expected to have different output now also?


<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/sql/tenant_creation.go:
```diff
@@ -390,19 +390,33 @@ func CreateTenantRecord(
-	}), tenantSpanConfig)
+	})
+
+	records, err := spanConfigs.GetSpanConfigRecords(ctx, []spanconfig.Target{endRecordTarget})
```
We're making some assumptions here for code that currently meets the assumptions but at a distance. Specifically we're assuming that the SQLTranslator's first emitted record, once spanconfig.Reconciler kicks in, starts at `keys.MakeTenantPrefix(tenantID)`. So it's queryable here. That record must always stay around -- the spanconfig.{Reconciler,Translator} can't get rid of the record with start key == `keys.MakeTenantPrefix(tenantID)`. If it did, then this read txn would not find anything and insert its own. Document the specifics will help, and we should, but it's worth adding an assertion in the spanconfig package to make sure we don't regress on this bug accidentally.

irfansharif replied:
> Could you point me to where to add the assertion? I'm not very familiar with this code.

@arulajmani is your resident expert (especially if this assertion is better suited within spanconfig.SQLTranslator), but one place to do it is here: https://github.com/cockroachdb/cockroach/blob/2f811420a1492ca547659003f193233341b7fd87/pkg/spanconfig/spanconfigreconciler/reconciler.go#L301

If we're running that code as a tenant, the storeWithLatestSpanConfigs can be queried/iterated over for the first span config, to sanity check that it's the one we expect -- starting at the tenant prefix. (For the system tenant the first span config will be over `/{Min-System/NodeLiveness}`, hence the check that we're not the system tenant).

irfansharif commented on pkg/sql/tenant_creation.go:
```diff
@@ -390,19 +390,33 @@ func CreateTenantRecord(
+	// tenant was already created then this tenant's endKey already exists.
+	// Skip creating this tenant's endKey if this already happened.
+	// See: https://github.com/cockroachdb/cockroach/issues/95882
+	toUpsert := []spanconfig.Record{startRecord}
```
Move this initialization of `toUpsert` above this whole code block that's trying to concern itself with the end key/end split. We already know we want the start record when initializing the tenant.

irfansharif reviewed (dismissed):
Left suggestions on commentary; do use the commentary to improve the commit message too perhaps. The code changes look right to me.

irfansharif reviewed (commented):
> Reviewed 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96014)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @knz)

___
*[`pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split_regression` line 127 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMjCszqCQmcExI3bHI2:-NMjFRo35fTDDuIm7cq3:bgwb1eg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fffb1bbe7ca04f5a4afc5ddecc415f92b3d453/pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split_regression#L127)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Is this expected to have different output now also?
</blockquote></details>

No, this looks correct.


<!-- Sent from Reviewable.io -->

ecwall commented on pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split_regression:
```diff
@@ -0,0 +1,128 @@
+# BUG: Show how creation of a tenant can end up overwriting span config state
```
Updated.

arulajmani reviewed (commented):
Let's file a separate issue about the bug Irfan pointed out as well? That seems legit -- I think it's a bit more general than the last tenant however. What we're really looking for is no tenant's last range should straddle tenant boundaries. If tenant n is dropped, and it deletes the split at the start of its keyspace, we'll be back to pre #92072 days for tenant n - 1. 

The easiest fix there would be to leave the single-key span config record at the start of tenant n's keyspace around when it gets dropped. @irfansharif, thoughts? 

> Reviewed 1 of 1 files at r1, 1 of 2 files at r2, 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-:-NMouUra-QNyuQzomBti:bfc4rqv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @irfansharif, and @knz)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMoniMt9CoqK3wA_qje):*
Let's not use imperative phrasing here. See https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#GitCommitMessages-Committitle.

Separately, consider pulling both these commits into one -- especially given you're making changes to the test from the first commit to the second, I don't see a good reason to have 2 separate commits here.  
___
*[`-- commits` line 14 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMooTbA3ol9P_GSHeRE):*
Same comment as above, let's not use the imperative form here. 
___
*[`pkg/sql/tenant_creation.go` line 399 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMjBXfqsAIXuG_SNZTd-r2-399:-NMotFJV6arTumcR0t3Z:bbblxvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fffb1bbe7ca04f5a4afc5ddecc415f92b3d453/pkg/sql/tenant_creation.go#L399)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

> Could you point me to where to add the assertion? I'm not very familiar with this code.

@arulajmani is your resident expert (especially if this assertion is better suited within spanconfig.SQLTranslator), but one place to do it is here: https://github.com/cockroachdb/cockroach/blob/2f811420a1492ca547659003f193233341b7fd87/pkg/spanconfig/spanconfigreconciler/reconciler.go#L301

If we're running that code as a tenant, the storeWithLatestSpanConfigs can be queried/iterated over for the first span config, to sanity check that it's the one we expect -- starting at the tenant prefix. (For the system tenant the first span config will be over `/{Min-System/NodeLiveness}`, hence the check that we're not the system tenant).
</blockquote></details>

I have no strong feelings about putting this in the Translator vs. the Reconciler. The line Irfan linked feels like a good place to put the assertion. 
___
*[`pkg/sql/tenant_creation.go` line 395 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMoumKSAC-apnkC_VQB:-NMoumKSAC-apnkC_VQC:bmrgzxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b032a638126db3bbc1e1d1e1d3d8818c8e16a92/pkg/sql/tenant_creation.go#L395)):*
> ```Go
> 	toUpsert := []spanconfig.Record{startRecord}
> 
> 	// Ensure we have a split at the current tenant's endKey in one of two ways:
> ```

nit: maybe this commentary will be clearer if you talk about tenants as tenant n, tenant n + 1?
___
*[`pkg/sql/tenant_creation.go` line 399 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMorgMaDjJuIWycn0y0:-NMorgMaDjJuIWycn0y1:b-aztfnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b032a638126db3bbc1e1d1e1d3d8818c8e16a92/pkg/sql/tenant_creation.go#L399)):*
> ```Go
> 	//    already exists and created this split at its own startKey when it was
> 	//    created.
> 	// 2) The next tenant does not exist so split now at the endKey.
> ```

Let's add some words about what happens when the next tenant is created in the future. Something about the new tenant overwriting this record during its creation, which will end up being a no-op, and continue to provide a split point at the end of the tenant's keyspace for this tenant. 
___
*[`pkg/sql/tenant_creation.go` line 402 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMos0UG5SNN0RaCY_Hh:-NMos0UG5SNN0RaCY_Hi:byxuvrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b032a638126db3bbc1e1d1e1d3d8818c8e16a92/pkg/sql/tenant_creation.go#L402)):*
> ```Go
> 	//
> 	// If this endKey split is not created via (1) or (2), the current tenant's
> 	// last range will have an endKey of /Max which is outside of it's keyspace.
> ```

That's only true if this is the last tenant, by ID, that exists. Let's drop the /Max phrasing -- instead, let's say it'll be outside of its keyspace. 

nit: s/it's/its
___
*[`pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split_regression` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMonwAMBSChOmpRSCqZ:-NMonwAMBSChOmpRSCq_:b-rglfka) ([raw file](https://github.com/cockroachdb/cockroach/blob/6551002e5493913a46ea10a307a7f46d8e5c395e/pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split_regression#L1)):*
> ```
> # BUG: Show how creation of a tenant can end up overwriting span config state
> ```

nit: "s/BUG/This test serves as a regression for <insert github link>"


<!-- Sent from Reviewable.io -->

ecwall committed:
multitenant: prevent tenant from overriding next tenant's span config

Fixes #95882

We want to ensure we have a split at the non-inclusive end of the tenant's
keyspace, which also happens to be the inclusive start of the next
tenant's (with ID=ours+1). If we didn't do anything here, and we were the
tenant with the highest ID thus far, our last range would extend to /Max.
If a tenant with a higher ID was created, when installing its initial span
config record, it would carve itself off (and our last range would only
extend to that next tenant's boundary), but this is a bit awkward for code
that wants to rely on the invariant that ranges for a tenant only extend
to the tenant's well-defined end key.

So how do we ensure this split at this tenant's non-inclusive end key?
Hard splits are controlled by the start keys on span config records[^1],
so we'll try insert one accordingly. But we cannot do this blindly. We
cannot assume that tenants are created in ID order, so it's possible that
the tenant with the next ID is already present + running. If so, it may
already have span config records that start at the key at which we want
to write a span config record for. Over-writing it blindly would be a
mistake -- there's no telling what config that next tenant has associated
for that span. So we'll do something simple -- we'll check transactionally
whether there's anything written already, and if so, do nothing. We
already have the split we need.

[^1]: See ComputeSplitKey in spanconfig.StoreReader.

Release note: None

ecwall reviewed (commented):
Created https://github.com/cockroachdb/cockroach/issues/96247

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-:-NN3wz9JBWCOZIpJxU12:bnyir52)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @knz)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMoniMt9CoqK3wA_qje):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's not use imperative phrasing here. See https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#GitCommitMessages-Committitle.

Separately, consider pulling both these commits into one -- especially given you're making changes to the test from the first commit to the second, I don't see a good reason to have 2 separate commits here.  
</blockquote></details>

Removed this comment and message.
___
*[`-- commits` line 14 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMooTbA3ol9P_GSHeRE):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Same comment as above, let's not use the imperative form here. 
</blockquote></details>

Fixed.
___
*[`pkg/sql/tenant_creation.go` line 399 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMjBXfqsAIXuG_SNZTd-r2-399:-NN3wKGnCAycVRZLsOtZ:b-d8mp32) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fffb1bbe7ca04f5a4afc5ddecc415f92b3d453/pkg/sql/tenant_creation.go#L399)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I have no strong feelings about putting this in the Translator vs. the Reconciler. The line Irfan linked feels like a good place to put the assertion. 
</blockquote></details>

Added a check here.
___
*[`pkg/sql/tenant_creation.go` line 395 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMoumKSAC-apnkC_VQB:-NN3oCj22-xY7k1cLROD:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b032a638126db3bbc1e1d1e1d3d8818c8e16a92/pkg/sql/tenant_creation.go#L395)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: maybe this commentary will be clearer if you talk about tenants as tenant n, tenant n + 1?
</blockquote></details>

Updated.



_Code quote:_
```Go
// Ensure we have a split at the current tenant's endKey in o
```
___
*[`pkg/sql/tenant_creation.go` line 399 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMorgMaDjJuIWycn0y0:-NN3p7E99aKMmzOocf9j:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b032a638126db3bbc1e1d1e1d3d8818c8e16a92/pkg/sql/tenant_creation.go#L399)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's add some words about what happens when the next tenant is created in the future. Something about the new tenant overwriting this record during its creation, which will end up being a no-op, and continue to provide a split point at the end of the tenant's keyspace for this tenant. 
</blockquote></details>

Updated.
___
*[`pkg/sql/tenant_creation.go` line 402 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMos0UG5SNN0RaCY_Hh:-NN3oM4s9zmBdAJ3ak6A:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b032a638126db3bbc1e1d1e1d3d8818c8e16a92/pkg/sql/tenant_creation.go#L402)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

That's only true if this is the last tenant, by ID, that exists. Let's drop the /Max phrasing -- instead, let's say it'll be outside of its keyspace. 

nit: s/it's/its
</blockquote></details>

Updated.
___
*[`pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split` line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMjqmhnYG0VK2NrPERN-r3-117:-NN3lX-G8QA5FfJibAqa:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b032a638126db3bbc1e1d1e1d3d8818c8e16a92/pkg/ccl/spanconfigccl/spanconfigreconcilerccl/testdata/multitenant/tenant_end_key_split#L117)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

This block can be removed since there's no mutation to speak of after the bug fix.
</blockquote></details>

Updated.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> The easiest fix there would be to leave the single-key span config record at the start of tenant n's keyspace around when it gets dropped. @irfansharif, thoughts?

If many tenants get deleted over time, over time we'll end up fragmenting the keyspace with empty ranges. I recommend solving this fully by ensuring that the "boundary" span configs are only left for the tenants that need it. i.e. If we're deleting tenant=11 and tenant=10 is around, leave the config at the start of tenant=11's keyspace. If tenant=12 then gets deleted, since there's no tenant=11 around to rely on that boundary, we don't need to keep it around.

> Reviewed 1 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-:-NN3xBaVBegCLaCoMPFo:b5etpcz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @knz)

___
*[`-- commits` line 14 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMooTbA3ol9P_GSHeRE):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Fixed.
</blockquote></details>

Thanks for updating the code comment elsewhere. Do use that here in this commit message too since its an improvement and clearer to readers.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks for taking the time to clean up the comments in `sql.CreateTenantRecord`. Like these subtle bugs have uncovered, this stuff can get tricky and the code here makes various assumptions. I have no doubt that taking the time to document this stuff will go a long way for future readers who come across this code! 

> If many tenants get deleted over time, over time we'll end up fragmenting the keyspace with empty ranges. I recommend solving this fully by ensuring that the "boundary" span configs are only left for the tenants that need it. i.e. If we're deleting tenant=11 and tenant=10 is around, leave the config at the start of tenant=11's keyspace. If tenant=12 then gets deleted, since there's no tenant=11 around to rely on that boundary, we don't need to keep it around.

That makes sense. I'd personally be happy solving the problem at hand without worrying about fragmenting the keyspace with empty ranges to start off with. We can always write some sort of de-fragmenting builtin after the fact, or whatever else we may want to do here. 

> Reviewed 1 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-:-NN8y4dTC4K2ZcHR9Xkr:bhszbg3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ecwall, @irfansharif, and @knz)

___
*[`-- commits` line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NN8sb7L0Ge-7PbCvIVk):*
"This patch ensures ... "
___
*[`pkg/spanconfig/spanconfigreconciler/reconciler.go` line 304 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NN8vAZT6r3fWxUa4wMx:-NN8vAZT6r3fWxUa4wMy:b-wgna0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8eee4dd39c5fa7139e266f69f51ade692b7776/pkg/spanconfig/spanconfigreconciler/reconciler.go#L304)):*
> ```Go
> 	if !f.codec.ForSystemTenant() {
> 		tenantPrefixKey := f.codec.TenantPrefix()
> 		// For secondary tenants, we assume that the SQLTranslator's first emitted
> ```

nit: Consider rewording this as follows:

```
We want to ensure tenant ranges do not straddle tenant boundaries. As such, a full reconciliation should always include a SpanConfig where the start key is keys.MakeTenantPrefix(tenantID). This ensures there is a split point right at the start of the tenant's keyspace, so that the last range of the previous tenant doesn't straddle across into this tenant. Also, sql.CreateTenantRecord relies on such a SpanConfigs existence to ensure the same thing for newly created tenants. 
```
___
*[`pkg/sql/tenant_creation.go` line 401 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NN8t5_Y9ExhMV0odWbY:-NN8t5_Y9ExhMV0odWbZ:bg07eex) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8eee4dd39c5fa7139e266f69f51ade692b7776/pkg/sql/tenant_creation.go#L401)):*
> ```Go
> 	// higher ID was created, when installing its initial span config record, it would
> 	// carve itself off (and our last range would only extend to that next tenant's
> 	// boundary), but this is a bit awkward for code that want to rely on the
> ```

"for code that wants"
___
*[`pkg/sql/tenant_creation.go` line 406 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NN8tTj1Dmxj6gUunzLt:-NN8tTj1Dmxj6gUunzLu:b-9by9me) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8eee4dd39c5fa7139e266f69f51ade692b7776/pkg/sql/tenant_creation.go#L406)):*
> ```Go
> 	//
> 	// So how do we ensure this split at this tenant's non-inclusive end key? Hard
> 	// splits are controlled by the start keys found span config records[^1], so we'll
> ```

"are controlled by the start keys on span config records"


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96014)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani, @irfansharif, and @knz)

___
*[`-- commits` line 14 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NMooTbA3ol9P_GSHeRE):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Thanks for updating the code comment elsewhere. Do use that here in this commit message too since its an improvement and clearer to readers.
</blockquote></details>

Moved here also
___
*[`-- commits` line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NN8sb7L0Ge-7PbCvIVk):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

"This patch ensures ... "
</blockquote></details>

Fixed
___
*[`pkg/spanconfig/spanconfigreconciler/reconciler.go` line 304 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NN8vAZT6r3fWxUa4wMx:-NNg-0xODoIvO4G_HC3O:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8eee4dd39c5fa7139e266f69f51ade692b7776/pkg/spanconfig/spanconfigreconciler/reconciler.go#L304)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Consider rewording this as follows:

```
We want to ensure tenant ranges do not straddle tenant boundaries. As such, a full reconciliation should always include a SpanConfig where the start key is keys.MakeTenantPrefix(tenantID). This ensures there is a split point right at the start of the tenant's keyspace, so that the last range of the previous tenant doesn't straddle across into this tenant. Also, sql.CreateTenantRecord relies on such a SpanConfigs existence to ensure the same thing for newly created tenants. 
```
</blockquote></details>

Updated.
___
*[`pkg/sql/tenant_creation.go` line 401 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NN8t5_Y9ExhMV0odWbY:-NNg-CoT0Pi0SJI7U_xY:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8eee4dd39c5fa7139e266f69f51ade692b7776/pkg/sql/tenant_creation.go#L401)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

"for code that wants"
</blockquote></details>

Fixed.
___
*[`pkg/sql/tenant_creation.go` line 406 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96014#-NN8tTj1Dmxj6gUunzLt:-NNg-U3L71O2pgXKAca-:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8eee4dd39c5fa7139e266f69f51ade692b7776/pkg/sql/tenant_creation.go#L406)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

"are controlled by the start keys on span config records"
</blockquote></details>

Fixed


<!-- Sent from Reviewable.io -->

ecwall commented:
bors r=arulajmani
